### ISSUE TITLE
Proposal: ECS Provider - Deprecate legacy ServerGroup configuration properties

### ISSUE THREAD

**Issue by pkandasamy91** [spinnaker/spinnaker#5827]
**Overview**

This issue is a proposal to deprecate and remove legacy ServerGroup configuration properties from the ECS configure server group wizard.

**Background**

In Spinnaker Release v1.18.0, the ECS provider started [supporting routing traffic from multiple load balancers to a single service](https://www.spinnaker.io/community/releases/versions/1-18-0-changelog#support-for-ecs-services-with-multiple-target-groups). In order to support this feature, a new target group mappings property was introduced, which allowed users to specify multiple target group mappings rather than the previous experience of one. The existing target group and load balanced container properties were maintained so as not to break existing customers who relied on those properties to deploy their services.

Currently, complex logic is written to support both legacy target group and updated target group mappings properties for ECS server group configurations. This has lead to issues in which server group configuration logic is not updated to support legacy properties when new features are released. This has lead to issues being raised, and for users to have a negative experience within the ECS provider.

**Issues caused by duplicate properties:**

* [Issue 5592](https://github.com/spinnaker/spinnaker/issues/5592)
* [PR 8224](https://github.com/spinnaker/deck/pull/8224)

**Fields Impacted**

As more features are added, maintaining logic to support both the legacy properties as well as the newly defined target group mapping properties adds unnecessary complexity for users. We propose deprecating the following properties from Clouddriver in Spinnaker Release v1.21.0, as we believe that it will reduce the complexity of configuring server groups by using the new target group mappings:

* *CreateServerGroupDescription.containerPort*
* *CreateServerGroupDescription.loadBalancedContainer*
* *CreateServerGroupDescription.targetGroup*

**Timeline**

This issue will remain open until the change is merged in, and we are targeting deprecating legacy server group properties in Spinnaker Release v1.21.0.  After that, newly added service features will not be available when using the legacy server group properties. This will be followed by removing logic using the deprecated properties in Spinnaker Release v1.23.0 and returning errors if the legacy properties are used.

Each milestone should be completed by the time the corresponding Spinnaker version is released.

Spinnaker Version | Release Window Opens | Milestone
-- | -- | --
1.18.0 (previous) | 1/22/2020 | Enabled support for multiple target groups for ECS server groups. Updated the UI for configuring target groups within a server group.
1.21.0 | 6/23/2020 | No longer support new features using legacy server group properties. (Deprecate legacy server group properties)
1.23.0 | 10/13/2020 | Remove logic using legacy server group properties and throw validation exceptions if they're still used (remove all existing logic using legacy server group properties and raise exceptions)

**Migrating from the previously used properties to the new target group mappings properties:**

There are two methods for migrating an existing service configured in Spinnaker: automatically update the existing configuration using the Deck UI, or manually update the pipeline configuration json.

**1) Automatically update the existing configuration using the Deck UI:**

* Navigate to the Pipelines tab of your Application and Configure your existing pipeline.
<img width="920" alt="Screen Shot 2020-05-12 at 12 49 18 PM" src="https://user-images.githubusercontent.com/44981951/84208786-5e622600-aa69-11ea-9985-e4c4ae0eaa2d.png">

* On the pipeline configuration page, edit your existing server group.
<img width="1051" alt="Screen Shot 2020-05-12 at 1 00 54 PM" src="https://user-images.githubusercontent.com/44981951/84208834-7afe5e00-aa69-11ea-9e82-e84a7340149a.png">

* In the Server Group Configuration Wizard, verify your target group mappings are correct.
    * If your container is configured using a task definition artifact:
<img width="708" alt="Screen Shot 2020-05-12 at 1 11 05 PM" src="https://user-images.githubusercontent.com/44981951/84208893-9e290d80-aa69-11ea-9455-4a0962fc023b.png">

    * If your container is defined using inputs:
<img width="813" alt="Screen Shot 2020-05-12 at 1 07 35 PM" src="https://user-images.githubusercontent.com/44981951/84208927-b436ce00-aa69-11ea-9395-d42008253de4.png">

    * Click Done to save the updated changes.

**2) Manually update the existing pipeline json:**

* An existing pipeline configuration json might look something like this:
```
{
    “containerPort“: 80,
    “copySourceScalingPoliciesAndActions": false,
    "delayBeforeDisableSec": 0,
    "delayBeforeScaleDownSec": 0,
    "dockerImageCredentialsSecret": "None (No registry credentials)",
    "dockerLabels": {},
    “loadBalancedContainer”: “name-of-my-container”,
    "securityGroups": [],
    "serviceDiscoveryAssociations": [],
    "stack": "prod",
    "strategy": "redblack",
    "subnetType": "private-subnet",
    "tags": {},
    "targetGroup": "my-target-group"
}
```

* Replace the following properties: 

```
“containerPort“: 80,
“loadBalancedContainer”: “name-of-my-container”,
"targetGroup": "my-target-group",

```
with the following new properties:

```
"targetGroupMappings": [{
    "containerName": "name-of-my-container",
    "containerPort": 80,
    "targetGroup": "my-target-group"
}],

```
* The new pipeline configuration might look something like:

```
{
    “copySourceScalingPoliciesAndActions": false,
    "delayBeforeDisableSec": 0,
    "delayBeforeScaleDownSec": 0,
    "dockerImageCredentialsSecret": "None (No registry credentials)",
    "dockerLabels": {},
    "securityGroups": [],
    "serviceDiscoveryAssociations": [],
    "stack": "prod",
    "strategy": "redblack",
    "subnetType": "private-subnet",
    "tags": {},
    "targetGroupMappings": [{
        "containerName": "name-of-my-container",
        "containerPort": 80,
        "targetGroup": "my-target-group"
    }],
}
```

**Comment by atyutyunnik** [comment]
This is how pipelines expressions look like, when used in my pipelines (provides a lot of flexibility):
<img width="721" alt="deploy_with_pipeline_expressions" src="https://user-images.githubusercontent.com/45210095/84335269-5974a380-ab49-11ea-9a13-bc4894deaf67.PNG">



The target group looks like this:
```
"targetGroup": "auth-${parameters.Environment}-service-tg"
```
but UI cannot properly render it (which is fine as long as it works):
<img width="277" alt="tg" src="https://user-images.githubusercontent.com/45210095/84335848-b1f87080-ab4a-11ea-92a7-d881e068f980.PNG">

**Comment by pkandasamy91** [comment]
@atyutyunnik Thanks for taking the time to provide me with additional information and screenshots! 

I believe there's a fairly straightforward way of updating the JSON directly in deck. I've added screenshots below (as well as in migration path 2 of the description). In your case, the values of each will be pipeline expressions, rather than static values. Please let me know if you're able to successfully (or unsuccessfully) update your deployment configuration.

* Navigate to the Pipelines tab of your Application and Configure your existing pipeline.
<img width="920" alt="Screen Shot 2020-05-12 at 12 49 18 PM" src="https://user-images.githubusercontent.com/44981951/84208786-5e622600-aa69-11ea-9985-e4c4ae0eaa2d.png">

* Edit your Deployment Configuration and select Edit stage as JSON*
<img width="985" alt="Screen Shot 2020-06-10 at 8 05 25 PM" src="https://user-images.githubusercontent.com/44981951/84340613-f76e6b00-ab55-11ea-8f68-7df11c05a505.png">

* Replace the following properties: 

```
“containerPort“: 80,
“loadBalancedContainer”: “name-of-my-container”,
"targetGroup": "my-target-group",

```
with the following new properties:

```
"targetGroupMappings": [{
    "containerName": "name-of-my-container",
    "containerPort": 80,
    "targetGroup": "my-target-group"
}],

```

* Maintaining duplicate properties (legacy as well as the new targetGroupMappings properties) should not cause you any issues. Feel free to remove the legacy properties once you confirm that everything continues to work.

**Comment by atyutyunnik** [comment]
Now imagine doing this for hundreds of pipelines... All manually...

**Comment by pkandasamy91** [comment]
> Now imagine doing this for hundreds of pipelines... All manually...

I completely understand and realize it's quite a bit of work, but it'll be a one time cost. I don't believe it's realistic to indefinitely support new features using both the new properties as well as the legacy properties. 

Do you think it would be reasonable if we only deprecated the legacy properties (no new features will be supported, but you can still use them as is) and did not remove the existing logic to throw an exception?

**Comment by atyutyunnik** [comment]
no, it's not one time cost and it's not quite a bit of work. It's a hundred times change and it's quite a lot of work.

Best recourse, IMHO, is retire the old logic eventually but provide a tool to detect/upgrade the pipelines, if the upgrade is trivial, as you say.

**Comment by pkandasamy91** [comment]
@atyutyunnik, thanks for taking the time to provide your feedback, we really appreciate it. I understand that this proposal does not address your concerns and as such, we'll hold off on making these changes. I'll provide an update once we determine an alternative approach.

**Comment by awsiv** [comment]
We use https://github.com/armory-io/terraform-provider-spinnaker  to maintain our spinnaker pipelines (using terraform modules) and automating pipeline updates at least for *target group mappings* should be fine

Not sure how complicated it would be for migrating existing server groups though

**Comment by xibz** [comment]
Who's working on this? It says it is in progress, but I am unsure of the progress here
